### ISSUE TITLE
Make use of similar open source project

### ISSUE THREAD

**Issue by Tarek-Hasan** [makehumancommunity/mpfb2#87]
Most of the Blender users have heard of or used MB-Lab (ManuelbastioniLAB) by Manuel Bastioni, one of the founder of MakeHuman project. MB-Lab is a Blender add-on for humanoid character creation. The original project was abandoned in 2018 and its fork was also unmaintained since 2021 which supports Blender 2.80. Still, it's the most feature rich open source character creation tool to this day, the only major drawbacks been lack of support for newer versions of Blender. Its features include

- Topology
   - Complete mesh with teeth, tongue, eyelashes, gum, and eyes as part of main mesh in order to morph them
   - Optimized topology for subdivision surface and sculpting.

- Parametric
   - Hundreds of morphing parameters.
   - Skin editor that automatically creates the displacement maps according to the character meta parameters (muscle, tone and age values).
   - Fuzzy logic algorithm for interpolation permits to mix the parameters in real-time.
   - Automatic generation of random characters with special constrains, to allow the possibility to keep constant the character's heritage.
    - Algorithm to modify and finish the morphing in order to fit the expressions to the face shape.

- Auto-modelling
  - Auto-modelling system to create characters starting from a 3d draft. The user just provides a very rough model of what he has in mind, and the system automatically turns it in a realistic and coherent character.
  - Measure system to create a character just giving the body measures. It supports 33 "classic" dimensions, as used in the most of anthropometric studies and researches.

- Anthropological phenotypes library
   The lab provides a library of preset that covers the most common human phenotypes, divided into four main classes: Caucasian, Asian, Afro and Latino. For each class, there is a specific set of phenotypes. Each phenotype can be loaded from the library and used as the base for a custom character, or mixed with another phenotype. The available phenotypes are:
   - Afro: Afromediterranean, Afroasian, Aboriginal, African.
   - Asian: Central Asian, North Asian, East Asian, South Asian, West Asian.
   - Caucasian: Central European, Afrocaucasian, East European, North European, Euromediterranean, Euroartic, North West European.
   - Latino: Central American, North American, South American.

- Create non-human characters: Anime, Realistic Anime, Elves and Dwarf.

- Advanced shaders
   - Supports Cycles and EEVEE
   - Procedural Eyeball and Iris shaders for Realistic Models
   - Shaders for skin, teeth, tongue, nail and eyelash
   - Texture maps for SSS, Roughness, Specular, Bump and Displacement
   - Cell shader for Anime models
   - Basic hair for human and anime, with advanced parametric shader.

- Rigging
   - Standard skeleton and a well tested rigging, developed in order to match most of external applications and motion capture files.
   - Auto fitting of skeleton.
   - Basic muscle system designed with the standard Blender bending bones, so that finalized characters can work even without the lab.
   - Inverse kinematic

- After Creation Tools
  - Along with a library of facial expressions, a custom rig is provided to help animate faces
  - Load new clothing and polygon hair onto characters
  - Proxy fitting to automatically fit with one-click clothes to the characters and their poses, without needing of vertgroups, weights or skeleton

I hope you can make use of the project to help improve MPFB2.
Repo of MB-Lab fork
https://github.com/animate1978/MB-Lab

**Comment by joepal1976** [comment]
TL;DR: MBL and MPFB2 are license incompatible. MBL technology cannot be used to improve MPFB2, since that would force MPFB2 into a license setup which is the opposite of what is desired.

For reference, the licenses are:

- MPFB: https://github.com/makehumancommunity/mpfb2/blob/master/LICENSE.md
- MBL: https://github.com/animate1978/MB-Lab/blob/master/license.txt

**Source code**
If a single line of code was taken from MBL, MPFB2 would need to change its license to AGPL. This might seem like a minor issue, but it has real-world consequences which might be hard to get an overview of.

Further, in MBL, the license is not limited to AGPL. It has further restrictions listed regarding output and renders. If any source code was taken from MBL, it would not be far-fetched to claim that these restrictions would apply to MPFB2 too.

For example: In MBL, an exported model is covered by AGPL. In MPFB2, no license claim is made regarding exports or output. Since all included assets in MPFB2 have already been released as CC0, the use of these do not impose any license restriction. 

**Assets**
This is where it really hurts. MBL uses AGPL as a license for assets too. The promise in MPFB2 is that all bundled assets, including the base mesh and morphs, are CC0. 

If you wanted to build your own commercial character generator, you could simply copy all the assets from MPFB2 without any fear of running into license issue, and without having to fulfill any obligations. If you wanted to reuse assets from MBL, your project would need to use AGPL as a license.

Theoretically, the real potential for re-use from MBL would be if the assets could be ported and merged into MPFB2. But since the MBL license states that both original assets and 3d output based on those assets are AGPL, MPFB2 would then no longer be able to promise that assets and output are free to use without any restrictions.

**In summary**
To be clear: I don't criticize the license setup of MBL. It is obviously perfectly valid to have a strong copyleft as a goal in an open source project. But as it stands, it is a diametrically opposite goal than that of MPFB2, where the goal is to be as permissive as possible. The only reason the source code license in MPFB2 is GPL, is because this is required by Blender. Otherwise, the MPFB2 license would be MIT or BSD.

So the problem isn't whether the features in MBL are good, available or desirable. The problem is that they can't be re-used for license reasons.

The sad state of things is that any feature in MBL you'd want in MPFB2 would need to be rewritten from scratch without reusing any existing code. This in order to avoid even the suspicion that MPFB2 assets or output might have the license restrictions imposed by MBL.


I am obviously aware of MBL, as I have been around since the dawn of time and during the split from MakeHuman. But for the reasons listed above, I have knowingly kept a hard separation between the projects.

**Comment by ashledombos** [comment]
It seems to be the same issue with CharMorph.

https://github.com/Upliner/CharMorph/blob/master/license.txt

**Comment by Tarek-Hasan** [comment]
> It seems to be the same issue with CharMorph.
> 
> https://github.com/Upliner/CharMorph/blob/master/license.txt

Not really!!
As stated in the license files of both MB-Lab and CharMorph (a rewrite of MB-Lab for improved performance and other features),
- "All python files released in the ManuelbastioniLAB package, are released under GNU General Public License 3."
- "All python files released in the CharMorph package, are released under GNU General Public License 3."

Meaning, the code is under GPL. But the assets (models and database) in both projects are under AGPL, as they both use the same assets from ManuelbastioniLAB. So I don't see any problem using code from any of these projects.

**Comment by joepal1976** [comment]
The issue description is too wide to act on. If there is a clearly delimited piece of functionality that you want ported, we can discuss the details about that specifically.
